src/VisualStudio/Core/Def/ProjectSystem/VisualStudioWorkspaceImpl.cs

```diff
@@ -1563,16 +1564,29 @@ internal void RemoveProjectFromMaps(CodeAnalysis.Project project)
 
     internal async Task RefreshProjectExistsUIContextForLanguageAsync(string language, CancellationToken cancellationToken)
     {
-        await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(alwaysYield: true, cancellationToken);
+        UIContext? uiContext;
 
-        var uiContext = _languageToProjectExistsUIContext.GetOrAdd(
-            language,
-            language => Services.GetLanguageServices(language).GetService<IProjectExistsUIContextProviderLanguageService>()?.GetUIContext());
+        using (_gate.DisposableWait(cancellationToken))
+        {
+            (uiContext, var isContextActive) = _languageToProjectExistsUIContext.GetOrAdd(
+                language,
+                language => (Services.GetLanguageServices(language).GetService<IProjectExistsUIContextProviderLanguageService>()?.GetUIContext(), false));
 
-        // UIContexts can be "zombied" if UIContexts aren't supported because we're in a command line build or in
-        // other scenarios.
-        if (uiContext == null || uiContext.IsZombie)
-            return;
+            // UIContexts can be "zombied" if UIContexts aren't supported because we're in a command line build or in
+            // other scenarios.
+            if (uiContext == null || uiContext.IsZombie)
+                return;
+
+            // Determine if there is a project with a matching language. Uses _projectSystemNameToProjectsMap as 
+            // that data structure is updated before calling into this method, whereas CurrentSolution may not be.
+            var projectExistsWithLanguage = _projectSystemNameToProjectsMap.Any(projects => projects.Value.Any(project => project.Language == language));
+            if (projectExistsWithLanguage == isContextActive)
+                return;
+
+            _languageToProjectExistsUIContext[language] = (uiContext, projectExistsWithLanguage);
+        }
+
+        await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(alwaysYield: true, cancellationToken);
 
         // Note: it's safe to read CurrentSolution here outside of any sort of lock.  We do all work here on the UI
```

Comment:
It might be worth rephrasing this comment to make it clear it's not using the isContextActive boolean that was inside the lock, since there's no guarantee that if one logical thread goes from false -> true and another goes from true -> false that this last bit will run in the "right order". But since you're _not_ using it...it's fine.

Reply:
Appended to the comment.